### PR TITLE
[feat] Remove deprecated features for 4.x

### LINKED ISSUE #1917: Remove deprecated `how` syntax in the `depends_on()` function

### LINKED ISSUE #1918: Remove support for deprecated version string syntax in `@required_version` decorator

### LINKED ISSUE #2017: Drop `rfm.*` and `sn.*` decorators that are now also exposed as class builtins.
All the decorators below need to go 🪚 in 4.0
- `rfm.run_before`
- `rfm.run_after`
- `rfm.require_deps`
- `rfm.final`
- `rfm.requred_version`
- `sn.sanity_function`

### PR DESCRIPTION
**Pull request by vkarak**
Closes #1917.
Closes #1918.
Closes #2017.

### CODE REVIEW

File: reframe/core/pipeline.py [hunk 1]
```diff
@@ -1821,15 +1734,13 @@ def by_part(src, dst):
             Passing an integer to the ``how`` argument as well as using the
             ``subdeps`` argument is deprecated.
 
+        .. versionchanged:: 4.0.0
+           Passing an integer to the ``how`` argument is no more supported.
```
Comment (jjotero) [inline reframe/core/pipeline.py:1738]: no longer supported?

**Review by jjotero — COMMENTED** [review]
Just a minor comment

**Review by ekouts — CHANGES_REQUESTED** [review]
1. There are some references to `rfm.run_after` in `reframe/core/launchers`,  `reframe/core/schedulers/__init__.py` and `reframe/utility/__init__.py` that still need to be changed. They are part of documentation.
2. I see a lot of tests using the `parameterized_test` decorator in the branch. These should be fixed in a different PR, right?
3. There are also many tests using the `@sn.sanity_function` decorator in the branch. Different PR as well?

**Review by jjotero — APPROVED** [review]

**Review by teojgo — APPROVED** [review]

**Comment by vkarak** [comment]
@ekouts I will fix (1). I guess that by saying "tests," you mean ReFrame tests, not unit tests. These should be fixed in different PRs.

**Review by ekouts — APPROVED** [review]
lgtm

**Pull request merged**